pilot/pkg/config/kube/ingress/servicestore.go

```diff
@@ -0,0 +1,123 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ingress
+
+import (
+	"sync"
+
+	knetworking "k8s.io/api/networking/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	"istio.io/istio/pkg/config"
+	"istio.io/istio/pkg/util/sets"
+)
+
+// serviceStore stores the services that are of port name type, specified by ingresses.
+type serviceStore struct {
```

Comment:
I don't think we need to store all of this, on service change we can list all Ingress in the namespace and compute it on the fly and enqueue them.

Reply:
I'm fine with this suggestion. But I just worry about the CPU cost in the case of that there are few ingress resources refer service by port name and most ingress resources just use the port number. Besides, the service store only records the port name type services refered by ingress resources, not all services. 

The `serviceStore` is an indexer used to help us quickly determine whether a service is referenced by the ingress using the port name. By this hepler store, we can reduce the cost of list ingress in every service change and make reflection of service change faster.

Looking forward to your review. Thanks.

Reply:
Yes, i agree with @howardjohn 

On service change, we can find the associated ingress, and pocess ingress after a debounce

Reply:
Actually, it is very uncommon case here

If we are storing all associated resources, i donot think istiod can really work on big scale clusters.

Reply:
> On service change, we can find the associated ingress, and pocess ingress after a debounce

Here the key is how to quickly find the associated ingress. We will do every list ingress operation in every service event  if we don't have a indexer like `serviceStore`. We just store the service name mapping to corresponding ingress  and it will not have so much RAM cost but we will get O(1) CPU cost.  Attention: here we only do save the services which is of port name type by referred by ingress, not all services.

> Actually, it is very uncommon case here

I agree that it is very uncommon case. So introducing `serviceStore` only handle this case will better than doing every list ingress operation in every service event.
We have many users who previous use the ingress-nginx that works well on the service port name. They feedbacked this issue to us when they choose to use ingress-istio.

If I'm wrong please tell me. Thanks.

Reply:
With a comparision to kubernetes service with targetPort name, the processing logic in endpoint_controller.go is similar.

It does not maintain a cache of service to pods mapping. It also runs in the fly.

Reply:
We may do more optimization by comparing the old service with the new service here, if the port changes, we do need to trigger ingress convertion

Reply:
Maintaining indexes like `serviceStore` has a high cost and very very high risk of bugs. See https://docs.google.com/document/d/1-ywpCnOfubqg7WAXSPf4YgbaFDBEU9HIqMWcxZLhzwE/edit#heading=h.ffjmk8byb9gt for more of my thoughts on the manner.

I think the List approach is good enough, and  importantly - very simple.

Reply:
Got it 👍.
The service store has already been removed.